### PR TITLE
feat(pod-delete): Add helper pod approach for getting the timestamp in pod-delete experiment

### PR DESCRIPTION
**Pull request by uditgaurav**
Signed-off-by: Udit Gaurav <udit.gaurav@mayadata.io>
**What this PR does / why we need it**:
- This PR contains the helper pod approach for running pod delete experiments. We are using the helper pod for keeping the track of total chaos duration (timestamp). Currently, all the Litmus generic experiments were using the helper pod approach to run some of its main chaos logic but pod delete was exceptional and wasn't using such approach. So this will create uniformity and also a much clear approach to run pod delete.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes: #1504

**Checklist**

* [x] Does this PR have a corresponding GitHub issue?
* [x] Have you included relevant README for the chaoslib/experiment with details?
* [x] Have you added debug messages where necessary? 
* [x] Have you added task comments where necessary? 
* [x] Have you tested the changes for possible failure conditions?
* [ ] Have you provided the positive & negative test logs for the litmusbook execution?
* [ ] Does the litmusbook ensure idempotency of cluster state?, i.e., is cluster restored to original state?
* [x] Have you used non-shell/command modules for Kubernetes tasks?
* [x] Have you (jinja) templatized custom scripts that is run by the litmusbook, if any? 
* [ ] Have you (jinja) templatized Kubernetes deployment manifests used by the litmusbook, if any?
* [ ] Have you reused/created util functions instead of repeating tasks in the litmusbook?
* [ ] Do the artifacts follow the appropriate directory structure? 
* [ ] Have you isolated storage (eg: OpenEBS) specific implementations, checks? 
* [ ] Have you isolated platform (eg: baremetal kubeadm/openshift/aws/gcloud) specific implementations, checks?  
* [x] Are the ansible facts well defined? Is the scope explicitly set for playbook & included utils?
* [x] Have you ensured minimum/careful usage of shell utilities (awk, grep, sed, cut, xargs etc.,)?
* [x] Can the limtusbook be executed both from within & outside a container (configurable paths, no hardcode)?
* [x] Can you suggest the minimal resource requirements for the litmusbook execution?
* [x] Does the litmusbook job artifact carry comments/default options/range for the ENV tunables?
* [x] Has the litmusbooks been linted? 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by uditgaurav** [comment]
- [Pod-delete-logs](https://notepad.pw/code/v0xyzwvj2) after the changes.

```bash
udit@ubuntu ~/g/s/g/l/litmus> kubectl get po -n litmus
NAME                                 READY   STATUS              RESTARTS   AGE
chaos-operator-ce-554d94dc74-d4fzt   1/1     Running             0          133m
nginx-7db9fccd9b-b8j78               0/1     ContainerCreating   0          1s
pod-delete-2ldvvr                    1/1     Running             0          27s
pod-delete-n6hjv-vgtwk               1/1     Running             0          43s
```

File: chaoslib/litmus/pod_delete/kill_random_pod.yml [hunk 1]
```diff
@@ -107,13 +97,29 @@
         retries: 90   
       when: "type is defined and type != ''"
 
-    when: "chaosDiffTimeStamp is not defined or (chaosDiffTimeStamp is defined and chaosDiffTimeStamp|int < c_duration|int)"
+    - name: "[Status]: Verify that the pod delete timestamp pod is running"
```
Comment (ksatchit) [inline chaoslib/litmus/pod_delete/kill_random_pod.yml:100]: Let us follow the same approach for termination as used in network-chaos and other experiments

File: chaoslib/litmus/pod_delete/pod-delete.j2 [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  name: pod-delete-{{ run_id }}
+  labels:
+    app: pod-delete
+    name: pod-delete-{{ run_id }}
+{% if chaos_uid is defined and chaos_uid != '' %}
+    chaosUID: {{ chaos_uid }}
+{% endif %}
+spec:
+  restartPolicy: Never
+  containers:
```
Comment (ksatchit) [inline chaoslib/litmus/pod_delete/pod-delete.j2:13]: Where is the chaos injection logic placed?

**Comment by uditgaurav** [comment]
```bash
ansible-playbook 2.7.3
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible-playbook
  python version = 2.7.15+ (default, Oct  7 2019, 17:39:04) [GCC 7.4.0]
Using /etc/ansible/ansible.cfg as config file
/etc/ansible/hosts did not meet host_list requirements, check plugin documentation if this is unexpected
/etc/ansible/hosts did not meet script requirements, check plugin documentation if this is unexpected
statically imported: /experiments/generic/pod_delete/pod_delete_ansible_prerequisites.yml

2020-05-21 22:24:55.291447 PLAYBOOK: pod_delete_ansible_logic.yml 
1 plays in ./experiments/generic/pod_delete/pod_delete_ansible_logic.yml

2020-05-21 22:24:55.308856 ************ BRACE YOURSELF, EXPERIMENT BEGINS! ************ 

2020-05-21 22:25:00.451824 Step: Gathering Facts 
task path: /experiments/generic/pod_delete/pod_delete_ansible_logic.yml:2
META: ran handlers

2020-05-21 22:25:00.948195 Step: [PreReq] Identify the chaos util for pod-delete experiment 
task path: /experiments/generic/pod_delete/pod_delete_ansible_prerequisites.yml:1
details => {"changed": true, "checksum": "a4b0c755a1d92a8ea0226d8000cc0d63b310e6a1", "dest": "/tmp/chaosutil.yml", "gid": 0, "group": "root", "md5sum": "868bdb8c466c56152cc0a1e5d2b7e5dc", "mode": "0644", "owner": "root", "size": 62, "src": "/tmp/ansible-tmp-1590099900.5-25356711580372/source", "state": "file", "uid": 0}

2020-05-21 22:25:01.010360 Step: [PreReq]: Including the chaos util for the pod-delete experiment 
task path: /experiments/generic/pod_delete/pod_delete_ansible_logic.yml:27
details => {"ansible_facts": {"c_util": "/chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml"}, "ansible_included_var_files": ["/tmp/chaosutil.yml"], "changed": false}
included: /utils/runtime/update_chaos_result_resource.yml for 127.0.0.1

2020-05-21 22:25:01.402222 Step: [PreReq]: Generate the chaos result CR to reflect SOT (Start of Test) 
task path: /utils/runtime/update_chaos_result_resource.yml:3
details => {"changed": true, "checksum": "cd4271e9e963f2977e802657322d92d2108da54d", "dest": "/tmp/chaos-result.yml", "gid": 0, "group": "root", "md5sum": "9238d1038be4b882caf101e5f2ea4f99", "mode": "0644", "owner": "root", "size": 256, "src": "/tmp/ansible-tmp-1590099901.17-67914166086978/source", "state": "file", "uid": 0}

2020-05-21 22:25:02.901375 Step: [PreReq]: Apply the chaos result CR for pod-delete experiment 
task path: /utils/runtime/update_chaos_result_resource.yml:13
details => {"changed": true, "cmd": "kubectl apply -f /tmp/chaos-result.yml -n litmus", "delta": "0:00:01.210452", "end": "2020-05-21 22:25:02.874632", "failed_when_result": false, "rc": 0, "start": "2020-05-21 22:25:01.664180", "stderr": "", "stderr_lines": [], "stdout": "chaosresult.litmuschaos.io/pod-delete unchanged", "stdout_lines": ["chaosresult.litmuschaos.io/pod-delete unchanged"]}

2020-05-21 22:25:03.041349 Step: [Info]: Display the application information passed via the test job 
task path: /experiments/generic/pod_delete/pod_delete_ansible_logic.yml:45
details => {
    "msg": [
        "The application info is as follows:", 
        "Namespace    : litmus", 
        "Label        : run=nginx", 
        "Ramp Time    : "
    ]
}
included: /utils/common/status_app_pod.yml for 127.0.0.1

2020-05-21 22:25:03.937973 Step: [Status]: Checking whether application pods are in running state 
task path: /utils/common/status_app_pod.yml:2
details => {"attempts": 1, "changed": true, "cmd": "kubectl get pods -n litmus -l run=nginx -o custom-columns=:.status.phase --no-headers", "delta": "0:00:00.662855", "end": "2020-05-21 22:25:03.911094", "rc": 0, "start": "2020-05-21 22:25:03.248239", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

2020-05-21 22:25:04.826101 Step: [Status]: Checking whether application containers are in running state 
task path: /utils/common/status_app_pod.yml:9
details => {"attempts": 1, "changed": true, "cmd": "kubectl get pod -n litmus -l run=nginx --no-headers -o jsonpath='{.items[*].status.containerStatuses[*].ready}' | tr ' ' '\\n' | uniq", "delta": "0:00:00.706816", "end": "2020-05-21 22:25:04.800855", "rc": 0, "start": "2020-05-21 22:25:04.094039", "stderr": "", "stderr_lines": [], "stdout": "true", "stdout_lines": ["true"]}
included: /chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml for 127.0.0.1

2020-05-21 22:25:05.022429 Step: [Prepare]: Deriving the chaos iterations 
task path: /chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml:1
details => {"ansible_facts": {"c_iterations": "3"}, "changed": false}

2020-05-21 22:25:05.143898 Step: [Prepare]: Deriving total chaos interval 
task path: /chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml:11
details => {"ansible_facts": {"total_c_interval": "15"}, "changed": false}

2020-05-21 22:25:05.997011 Step: [Prepare]: Generate a run id if not passed from the engine/experiment 
task path: /chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml:31
details => {"changed": true, "cmd": "echo $(mktemp) | cut -d '.' -f 2 | cut -c -6", "delta": "0:00:00.561980", "end": "2020-05-21 22:25:05.975286", "rc": 0, "start": "2020-05-21 22:25:05.413306", "stderr": "", "stderr_lines": [], "stdout": "bbzjVP", "stdout_lines": ["bbzjVP"]}

2020-05-21 22:25:06.056809 Step: set_fact 
task path: /chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml:35
details => {"ansible_facts": {"run_id": "bbzjvp"}, "changed": false}

2020-05-21 22:25:06.109970 Step: [Prepare]: Setting type to generic 
task path: /chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml:41
details => {"ansible_facts": {"type": "generic"}, "changed": false}

2020-05-21 22:25:06.162777 Step: [Prepare]: Setting the chaos_status to active 
task path: /chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml:47
details => {"ansible_facts": {"chaos_status": "active"}, "changed": false}

2020-05-21 22:25:06.455529 Step: [Prepare]: Patch the pod delete timestamp pod 
task path: /chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml:155
details => {"changed": true, "checksum": "b5a9cb4e0c62a0bbec66f4bacf11fb71b8d97c02", "dest": "/tmp/pod-delete.yml", "gid": 0, "group": "root", "md5sum": "05e75bd6395882c52c9dd04f04fbee4d", "mode": "0644", "owner": "root", "size": 1274, "src": "/tmp/ansible-tmp-1590099906.22-129923667882751/source", "state": "file", "uid": 0}

2020-05-21 22:25:07.435918 Step: [Prepare]: Creating timestamp pod for pod delete chaos 
task path: /chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml:160
details => {"changed": true, "cmd": "kubectl create -f /tmp/pod-delete.yml -n litmus", "delta": "0:00:00.783431", "end": "2020-05-21 22:25:07.408696", "rc": 0, "start": "2020-05-21 22:25:06.625265", "stderr": "", "stderr_lines": [], "stdout": "pod/pod-delete-bbzjvp created", "stdout_lines": ["pod/pod-delete-bbzjvp created"]}
FAILED - RETRYING: [Status]: Checking whether application pods are in completed state (15 retries left).
FAILED - RETRYING: [Status]: Checking whether application pods are in completed state (14 retries left).
FAILED - RETRYING: [Status]: Checking whether application pods are in completed state (13 retries left).
FAILED - RETRYING: [Status]: Checking whether application pods are in completed state (12 retries left).
FAILED - RETRYING: [Status]: Checking whether application pods are in completed state (11 retries left).
FAILED - RETRYING: [Status]: Checking whether application pods are in completed state (10 retries left).
FAILED - RETRYING: [Status]: Checking whether application pods are in completed state (9 retries left).
FAILED - RETRYING: [Status]: Checking whether application pods are in completed state (8 retries left).
FAILED - RETRYING: [Status]: Checking whether application pods are in completed state (7 retries left).
FAILED - RETRYING: [Status]: Checking whether application pods are in completed state (6 retries left).
FAILED - RETRYING: [Status]: Checking whether application pods are in completed state (5 retries left).
FAILED - RETRYING: [Status]: Checking whether application pods are in completed state (4 retries left).
FAILED - RETRYING: [Status]: Checking whether application pods are in completed state (3 retries left).
FAILED - RETRYING: [Status]: Checking whether application pods are in completed state (2 retries left).
FAILED - RETRYING: [Status]: Checking whether application pods are in completed state (1 retries left).

2020-05-21 22:25:37.042237 Step: [Status]: Checking whether application pods are in completed state 
task path: /chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml:172
fatal: [127.0.0.1]: FAILED! => {"attempts": 15, "changed": true, "cmd": "kubectl get pods -n litmus -l name=pod-delete-bbzjvp -o custom-columns=:.status.phase --no-headers", "delta": "0:00:00.714916", "end": "2020-05-21 22:25:37.015400", "rc": 0, "start": "2020-05-21 22:25:36.300484", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}
...ignoring

2020-05-21 22:25:37.992803 Step: [CleanUP]: Delete Pod delete timestamp pod 
task path: /chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml:180
details => {"changed": true, "cmd": "kubectl delete -f /tmp/pod-delete.yml -n litmus", "delta": "0:00:00.765716", "end": "2020-05-21 22:25:37.962682", "rc": 0, "start": "2020-05-21 22:25:37.196966", "stderr": "", "stderr_lines": [], "stdout": "pod \"pod-delete-bbzjvp\" deleted", "stdout_lines": ["pod \"pod-delete-bbzjvp\" deleted"]}

2020-05-21 22:25:38.062379 Step: [Mark]: Setting chaos_status to terminate 
task path: /chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml:187
details => {"ansible_facts": {"chaos_status": "terminate"}, "changed": false}
included: /utils/common/status_app_pod.yml for 127.0.0.1

2020-05-21 22:25:39.218106 Step: [Status]: Checking whether application pods are in running state 
task path: /utils/common/status_app_pod.yml:2
details => {"attempts": 1, "changed": true, "cmd": "kubectl get pods -n litmus -l run=nginx -o custom-columns=:.status.phase --no-headers", "delta": "0:00:00.724107", "end": "2020-05-21 22:25:39.191085", "rc": 0, "start": "2020-05-21 22:25:38.466978", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

2020-05-21 22:25:40.126886 Step: [Status]: Checking whether application containers are in running state 
task path: /utils/common/status_app_pod.yml:9
details => {"attempts": 1, "changed": true, "cmd": "kubectl get pod -n litmus -l run=nginx --no-headers -o jsonpath='{.items[*].status.containerStatuses[*].ready}' | tr ' ' '\\n' | uniq", "delta": "0:00:00.673839", "end": "2020-05-21 22:25:40.101050", "rc": 0, "start": "2020-05-21 22:25:39.427211", "stderr": "", "stderr_lines": [], "stdout": "true", "stdout_lines": ["true"]}

2020-05-21 22:25:40.224689 Step: set_fact 
task path: /experiments/generic/pod_delete/pod_delete_ansible_logic.yml:97
details => {"ansible_facts": {"flag": "Pass"}, "changed": false}

2020-05-21 22:25:40.284221 Step: [Result]: Getting the final result of pod-delete experiment 
task path: /experiments/generic/pod_delete/pod_delete_ansible_logic.yml:100
details => {
    "msg": "pod-delete experiment has been Passed"
}
included: /utils/runtime/getting_failure_step.yml for 127.0.0.1

2020-05-21 22:25:40.631134 Step: [Failure-Detection]: Recording the name of failed step 
task path: /utils/runtime/getting_failure_step.yml:44
details => {"ansible_facts": {"failStep": "N/A"}, "changed": false}
included: /utils/runtime/update_chaos_result_resource.yml for 127.0.0.1

2020-05-21 22:25:41.683327 Step: [Result]: Update the chaos result CR to reflect EOT (End of Test) 
task path: /utils/runtime/update_chaos_result_resource.yml:23
details => {"changed": true, "checksum": "7c5439829560fcaf86ccbc2a97357c2cab0d6d7e", "dest": "/tmp/chaos-result.yml", "gid": 0, "group": "root", "md5sum": "03b46f21f022d67bc96feb473cde7ba8", "mode": "0644", "owner": "root", "size": 275, "src": "/tmp/ansible-tmp-1590099940.88-224224775721233/source", "state": "file", "uid": 0}

2020-05-21 22:25:42.734895 Step: [The END]: Apply the chaos result CR for pod-delete experiment 
task path: /utils/runtime/update_chaos_result_resource.yml:33
details => {"changed": true, "cmd": "kubectl apply -f /tmp/chaos-result.yml -n litmus", "delta": "0:00:00.828923", "end": "2020-05-21 22:25:42.708002", "failed_when_result": false, "rc": 0, "start": "2020-05-21 22:25:41.879079", "stderr": "", "stderr_lines": [], "stdout": "chaosresult.litmuschaos.io/pod-delete configured", "stdout_lines": ["chaosresult.litmuschaos.io/pod-delete configured"]}
META: ran handlers
META: ran handlers

2020-05-21 22:25:42.739110 ************ RELAX, EXPERIMENT ENDS! ************ 
127.0.0.1                  : ok=32   changed=14   unreachable=0    failed=0 
```

File: chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml [hunk 3]
```diff
@@ -33,15 +42,41 @@
...
+  - name: "[Prepare]: Creating timestamp pod for pod delete chaos"
+    shell: kubectl create -f /tmp/pod-delete.yml -n {{ c_ns }}
+    args:
+      executable: /bin/bash
+    register: helper_pod
+
+  - name: "[Status]: Checking whether application pods are in completed state"
+    shell: kubectl get pods -n {{ c_ns }} -l name=pod-delete-{{ run_id }} -o custom-columns=:.status.phase --no-headers
+    register: result
+    ignore_errors: yes
```
Comment (ksatchit) [inline chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml:63]: can we use the chaos time diff to verify success or failure of the job? Refer the container kill exp. cc: @ispeakc0de

File: chaoslib/litmus/pod_delete/pod-delete.j2 [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  name: pod-delete-{{ run_id }}
+  labels:
+    app: pod-delete
+    name: pod-delete-{{ run_id }}
+{% if chaos_uid is defined and chaos_uid != '' %}
+    chaosUID: {{ chaos_uid }}
+{% endif %}
+spec:
+  restartPolicy: Never
+  serviceAccountName: pod-delete-sa
```
Comment (uditgaurav) [inline chaoslib/litmus/pod_delete/pod-delete.j2:13]: should it have custom support @ksatchit ?

Comment (ksatchit) [reply chaoslib/litmus/pod_delete/pod-delete.j2:13]: yes pls

File: chaoslib/litmus/pod_delete/pod-delete.j2 [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      - name: RAMP_TIME
+        value: "{{ ramp_time }}"
+{% endif %}
+    volumeMounts:
+      - name: dockersocket
+        mountPath: /var/run/docker.sock
+  volumes:
+    - hostPath:
+        path: /var/run/docker.sock
+      name: dockersocket
```
Comment (ksatchit) [inline chaoslib/litmus/pod_delete/pod-delete.j2:70]: Do we need docker socket mounted for this script?

File: chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml [hunk 6]
```diff
@@ -1,53 +1,110 @@
...
+    - name: "[Prepare]: Getting the chaos start timestamp"
+      set_fact: chaosStartTimeStamp="{{lookup('pipe','date \"+%Y-%m-%d %H:%M:%S\"')}}"
+
+    - name: "[Prepare]: Create helper pod for pod delete chaos"
+      shell: kubectl create -f /tmp/pod-delete-chaos.yml -n {{ c_ns }}
+      args:
+        executable: /bin/bash
+      register: helper_pod
+
+    - name: "[Status]: Checking whether application pods are in completed state"
```
Comment (ksatchit) [inline chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml:63]: helper pod (name)?

File: chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml [hunk 7]
```diff
@@ -1,53 +1,110 @@
...
+          args: 
+            executable: /bin/bash
+          register: result_status
+          until: result_status.stdout ==''
+          delay: 2
+          retries: 90
+
+      when: "chaosDiffTimeStamp|int >= c_duration|int"
+      
+    - fail:
```
Comment (ispeakc0de) [inline chaoslib/litmus/pod_delete/pod_failure_by_litmus.yml:98]: Can we move this task before the helper pod deletion?

**Pull request merged**